core/src/test/java/com/linecorp/armeria/client/HttpClientIntegrationTest.java

```diff
@@ -724,6 +734,39 @@ void testUpgradeRequestExecutesLogicOnlyOnce() throws Exception {
         clientFactory.close();
     }
 
+    @ParameterizedTest
+    @ValueSource(booleans = { true, false })
+    void requestAbortWithException(boolean isAbort) {
+        final HttpClient client = HttpClient.of(server.httpUri("/"));
+        final HttpRequestWriter request = HttpRequest.streaming(HttpMethod.GET, "/client-aborted");
+
+        final HttpResponse response = client.execute(request);
+        final IllegalStateException badState = new IllegalStateException("bad state");
+        if (isAbort) {
+            request.abort(badState);
+        } else {
+            request.close(badState);
+        }
+        // request cause is obtained immediately
+        assertThatThrownBy(() -> response.aggregate().join())
+                .isInstanceOf(CompletionException.class)
+                .hasCause(badState);
+    }
+
+    @Test
+    void responseAbortWithException() throws InterruptedException {
+        final HttpClient client = HttpClient.of(server.httpUri("/"));
+        final HttpRequest request = HttpRequest.streaming(HttpMethod.GET, "/client-aborted");
+        final HttpResponse response = client.execute(request);
+        await().untilAsserted(() -> assertThat(completed).isTrue());
```

Comment:
I guess this can be `await().until(() => completed)`. Though actually it might be a reason to stick with `AtomicReference` since there's also `await().untilTrue(AtomicReference)`